northstar/src/runtime/state.rs

```diff
@@ -109,615 +89,464 @@ impl ProcessContext {
-            let (dir, repository) = prepare_hello_world_repo(&name).await?;
+            let (dir, repository) = prepare_internal_repository(&name).await?;
             repositories.insert(name, repository);
             dir
         };
 
+        // Build a map of repositories from the configuration
         for (id, repository) in &config.repositories {
             let key = {
                 if let Some(key) = repository.key.as_ref() {
-                    Some(keys::load(&key).await.map_err(Error::Key)?)
+                    Some(key::load(&key).await.map_err(Error::Key)?)
                 } else {
                     None
                 }
             };
             let dir = repository.dir.clone();
-            repositories.insert(
-                id.clone(),
-                Repository {
-                    id: id.clone(),
-                    dir,
-                    key,
-                },
-            );
+            repositories.insert(id.clone(), Repository::new(id.clone(), dir, key).await?);
         }
 
-        let minijail = Minijail::new(tx.clone(), config).map_err(Error::Process)?;
+        // Build a list of npks from the repositories
+        let mut npks = HashMap::new();
+        for (id, repository) in &repositories {
+            for ((name, version), npk) in repository.npks() {
+                let key = ContainerKey::new(name.clone(), version.clone(), id.clone());
+                npks.insert(key, npk.clone());
+            }
+        }
 
-        let mut state = State {
-            events_tx: tx,
+        let minijail = Minijail::new(events_tx.clone(), config).map_err(Error::Process)?;
+        let mount_control = MountControl::new(&config).await.map_err(Error::Mount)?;
+
+        Ok(State {
+            events_tx,
             repositories,
-            applications: HashMap::new(),
-            resources: HashMap::new(),
+            npks,
+            containers: HashMap::new(),
             config,
             minijail,
-            mount_control: MountControl::new(&config).await.map_err(Error::Mount)?,
+            mount_control,
             #[cfg(debug_assertions)]
             internal_repository,
-        };
-
-        // Mount all the containers from each repository
-        state.mount_repositories().await?;
-        Ok(state)
+        })
     }
 
-    /// Return an iterator over all known applications
-    pub fn applications(&self) -> impl iter::Iterator<Item = &Application> {
-        self.applications.values()
-    }
+    /// Mount `key`
+    async fn mount(&self, key: &ContainerKey) -> Result<Container, Error> {
+        // Repository key
+        let repository_key = self
+            .repositories
+            .get(key.repository())
+            .and_then(|r| r.key.as_ref());
+        let npk = self
+            .npks
+            .get(&key)
+            .ok_or_else(|| Error::UnknownApplication(key.clone()))?;
+
+        // Load NPK
+        let npk = Npk::from_path(npk, repository_key)
```

Comment:
this is really misleading...it's either an npk or an npk-path. too bad rust allows for reassigning to the same identifier.

Reply:
It's type safe and within a scope unique. I love this feature :-)

Reply:
I know it is technically typesafe but still does not contribute to understandable code in some instances such as this. I would not use it here.